core/src/test/kotlin/fr/sncf/osrd/pathfinding/PathfindingElectrificationTest.kt

```diff
@@ -0,0 +1,198 @@
+package fr.sncf.osrd.pathfinding
+
+import fr.sncf.osrd.api.ApiTest
+import fr.sncf.osrd.api.TrackLocation
+import fr.sncf.osrd.api.pathfinding.IncompatibleConstraintsPathResponse
+import fr.sncf.osrd.api.pathfinding.NoPathFoundException
+import fr.sncf.osrd.api.pathfinding.PathfindingBlockSuccess
+import fr.sncf.osrd.api.pathfinding.runPathfinding
+import fr.sncf.osrd.sim_infra.api.NeutralSection
+import fr.sncf.osrd.train.TestTrains
+import fr.sncf.osrd.utils.Direction
+import fr.sncf.osrd.utils.DummyInfra
+import fr.sncf.osrd.utils.units.Offset
+import fr.sncf.osrd.utils.units.meters
+import org.assertj.core.api.Assertions.assertThat
+import org.assertj.core.api.Assertions.assertThatThrownBy
+import org.junit.jupiter.api.Assertions.assertNotEquals
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.TestInstance
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.Arguments
+import org.junit.jupiter.params.provider.MethodSource
+import java.util.stream.Stream
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class PathfindingElectrificationTest : ApiTest() {
+    @Test
+    fun incompatibleElectrificationsTest() {
+        /*        N
+                ^  \
+               /    v
+        a --> b     d --> e
+               \    ^
+                v  /
+                 S
+         */
+        val infra = DummyInfra()
+        infra.addBlock("a", "b")
+        infra.addBlock("b", "N")
+        infra.addBlock("b", "S")
+        infra.addBlock("N", "d")
+        infra.addBlock("S", "d")
+        infra.addBlock("d", "e")
+        val waypointsStart = listOf(TrackLocation("a->b", Offset.zero()))
+        val waypointsEnd = listOf(TrackLocation("d->e", Offset(100.meters)))
+
+        assert(TestTrains.FAST_ELECTRIC_TRAIN.modeNames.contains("25000V"))
+        for (block in infra.blockPool) block.voltage = "25000V"
+
+        // Run a pathfinding with all blocks electrified
+        val normalPathResp =
+            runPathfinding(
+                infra.fullInfra(),
+                getPathfindingBlockRequest(
+                    TestTrains.FAST_ELECTRIC_TRAIN,
+                    listOf(waypointsStart, waypointsEnd)
+                )
+            )
+        assertThat(normalPathResp).isExactlyInstanceOf(PathfindingBlockSuccess::class.java)
+
+        val normalTracks =
+            (normalPathResp as PathfindingBlockSuccess)
+                .trackSectionRanges
+                .map { it.trackSection }
+                .toList()
+
+        // Removes electrification in the section used by the normal train
+        val normalPathTrackAfterB = normalTracks.intersect(setOf("b->N", "b->S"))
+        assert(normalPathTrackAfterB.size == 1)
+        infra.blockPool.forEach {
+            if (it.name == normalPathTrackAfterB.first()) it.voltage = "bad_elec"
+        }
```

Comment:
```suggestion
        infra.blockPool.forEach {
            if (it.name == normalPathTrackAfterB.single()) it.voltage = "bad_elec"
        }
```

(`single` is equivalent to `assert size == 1` and `.first`)

Reply:
Done (and propagated) https://github.com/OpenRailAssociation/osrd/pull/12167/commits/46129c9bd1bddc9a2d7d5f97b465bacfd697ea26